src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,138 @@
+    }
+
+    private Object[] grow() {
+        int newCapacity = elementData.length * 3 / 2 + 1;
+        elementData = Arrays.copyOf(elementData, newCapacity);
+        return elementData;
+    }
+
+    public void checkIfElementExists(T element) {
+        boolean found = false;
+        for (int i = 0; i < size; i++) {
+            if ((element == null && elementData[i] == null)
+                    || (element != null && element.equals(elementData[i]))) {
+                found = true;
+                break;
+            }
+        }
+        if (!found) {
+            throw new NoSuchElementException("Element not found in the array.");
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        rangeCheck();
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bound");
+        }
+        rangeCheck();
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
-    public void addAll(List<T> list) {
-
+    public void addAll(List<T> o) {
+        for (int i = 0; i < o.size(); i++) {
+            this.add(o.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexCheck(index);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        indexCheck(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexCheck(index);
+        final T removedElement = (T) elementData[index];
+        System.arraycopy(elementData, index + 1, elementData, index, size - index - 1);
+        elementData[size - 1] = null;
+        size--;
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
+        checkIfElementExists(element);
+        for (int i = 0; i < size; i++) {
+            if ((element == null && elementData[i] == null)
+                    || (element != null && element.equals(elementData[i]))) {
+                T removedElement = remove(i);
+                return removedElement;
+            }
+        }
         return null;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("[");
+
+        for (int i = 0; i < size; i++) {
+            if (elementData[i] == null) {
+                return sb.append("]").toString();
+            }
+            if (elementData[i + 1] == null) {
+                return sb.append(elementData[i]).append("]").toString();
```

Comment:
The logic in the `toString` method is incorrect. The check `if (elementData[i] == null)` will prematurely terminate the string representation if any element is null, which may not be the intended behavior. Similarly, `if (elementData[i + 1] == null)` will cause an `ArrayIndexOutOfBoundsException` when `i` is the last index. Consider revising this logic to correctly iterate through all elements up to `size`.